File: app/Services/PartnerVenueVisitService.php
<?php
// Copyright © 2024–2026 Jasper Ford
// SPDX-License-Identifier: AGPL-3.0-or-later
// Author: Jasper Ford
// See NOTICE file for attribution and acknowledgements.

declare(strict_types=1);

namespace App\Services;

use App\Core\TenantContext;
use App\Models\PartnerMemberPass;
use App\Models\PartnerVenue;
use App\Models\PartnerVenueVisit;
use Illuminate\Database\UniqueConstraintViolationException;
use Illuminate\Support\Facades\Cache;
use Illuminate\Support\Facades\DB;

/**
 * PartnerVenueVisitService — member passes and the visit ledger.
 *
 * The flow deliberately mirrors the volunteering QR check-in that is already
 * proven in production: the member's pass QR encodes a FRONTEND URL, so venue
 * staff can scan it with any phone camera and land on a page that asks them to
 * confirm. Recording therefore always happens as an authenticated, authorised
 * staff action — a member cannot self-record by photographing their own pass.
 *
 * A visit records engagement only. No discount is calculated, applied or
 * settled here, and no money or time credit moves.
 */
class PartnerVenueVisitService
{
    public const ENGAGEMENT_ACTION = 'venue_visit';

    public function __construct(
        private readonly PartnerVenueService $venues,
    ) {}

    /**
     * Get (or lazily create) the caller's membership pass.
     *
     * @return array{token:string, qr_url:string, status:string, last_used_at:string|null}
     */
    public function getOrCreatePass(int $userId): array
    {
        $tenantId = TenantContext::getId();

        $pass = PartnerMemberPass::query()->where('user_id', $userId)->first();

        if ($pass === null) {
            // Serialise check-then-insert; two concurrent calls would otherwise
            // both find nothing and race on the (tenant_id, user_id) unique key.
            $lock = Cache::lock(sprintf('partner_member_pass:%d:%d', $tenantId, $userId), 10);

            if ($lock->get()) {
                try {
                    $pass = PartnerMemberPass::query()->where('user_id', $userId)->first();

                    if ($pass === null) {
                        $pass = new PartnerMemberPass([
                            'user_id' => $userId,
                            'token' => bin2hex(random_bytes(32)),
                            'status' => 'active',
                        ]);
                        $pass->tenant_id = $tenantId;
                        $pass->save();
                    }
                } finally {
                    $lock->release();
                }
            } else {
                // Lost the lock race — the winner has written the row by now.
                $pass = PartnerMemberPass::query()->where('user_id', $userId)->firstOrFail();
            }
        }

        if ($pass->status !== 'active') {
            $pass->status = 'active';
            $pass->save();
        }

        return [
            'token' => (string) $pass->token,
            'qr_url' => $this->passUrl((string) $pass->token),
            'status' => (string) $pass->status,
            'last_used_at' => $pass->last_used_at?->toIso8601String(),
        ];
    }

    /**
     * Rotate the caller's pass token, invalidating the old QR.
     *
     * @return array{token:string, qr_url:string, status:string, last_used_at:string|null}
     */
    public function rotatePass(int $userId): array
    {
        $this->getOrCreatePass($userId);

        $pass = PartnerMemberPass::query()->where('user_id', $userId)->firstOrFail();
        $pass->token = bin2hex(random_bytes(32));
        $pass->status = 'active';
        $pass->save();

        return [
            'token' => (string) $pass->token,
            'qr_url' => $this->passUrl((string) $pass->token),
            'status' => (string) $pass->status,
            'last_used_at' => $pass->last_used_at?->toIso8601String(),
        ];
    }

    public function passUrl(string $token): string
    {
        return TenantContext::getFrontendUrl()
            . TenantContext::getSlugPrefix()
            . '/venues/checkin/' . $token;
    }

    /**
     * Record a visit from a scanned member pass.
     *
     * Result statuses:
     *  - recorded              a new visit row was written
     *  - already_recorded_today the member was already recorded at this venue today
     *  - needs_venue           the staff member covers several venues; pick one
     *  - invalid_pass          unknown, revoked or cross-tenant token
     *  - forbidden             the caller is not staff at any eligible venue
     *
     * @return array<string, mixed>
     */
    public function recordVisit(string $token, int $staffUserId, ?int $venueId = null): array
    {
        $tenantId = TenantContext::getId();

        $pass = PartnerMemberPass::query()
            ->where('token', $token)
            ->where('status', 'active')
            ->first();

        if ($pass === null) {
            return ['status' => 'invalid_pass'];
        }

        $memberId = (int) $pass->user_id;

        $eligible = $this->venues->venuesForStaff($staffUserId);

        if ($eligible === []) {
            return ['status' => 'forbidden'];
        }

        if ($venueId !== null) {
            if (! in_array($venueId, $eligible, true)) {
                return ['status' => 'forbidden'];
            }
        } elseif (count($eligible) === 1) {
            $venueId = $eligible[0];
        } else {
            return [
                'status' => 'needs_venue',
                'venues' => PartnerVenue::query()
                    ->whereIn('id', $eligible)
                    ->orderBy('name')
                    ->get(['id', 'name'])
                    ->map(static fn ($venue): array => [
                        'id' => (int) $venue->id,
                        'name' => (string) $venue->name,
                    ])
                    ->all(),
            ];
        }

        $venue = PartnerVenue::query()->find($venueId);

        if ($venue === null || $venue->status !== 'active') {
            return ['status' => 'forbidden'];
        }

        if ($memberId === $staffUserId) {
            // Staff recording their own visit would make the ledger
            // self-attested; the same posture as volunteering's
            // self-verification block.
            return ['status' => 'forbidden'];
        }

        $today = now()->toDateString();
        $created = true;

        try {
            $visit = new PartnerVenueVisit([
                'venue_id' => (int) $venue->id,
                'user_id' => $memberId,
                'recorded_by_user_id' => $staffUserId,
                'source' => 'member_pass',
                'visited_on' => $today,
                'visited_at' => now(),
            ]);
            $visit->tenant_id = $tenantId;
            $visit->save();
        } catch (UniqueConstraintViolationException) {
            // Same member, same venue, same day — the daily unique key makes a
            // rescan a no-op rather than an error.
            $created = false;
        }

        $pass->last_used_at = now();
        $pass->save();

        $engagement = ['xp_awarded' => 0, 'completed_challenges' => []];

        if ($created) {
            $engagement = EngagementService::record(
                $memberId,
                self::ENGAGEMENT_ACTION,
                'partner_venue:' . $venue->id . ':' . $today,
                __('api.partner_venue_visit_xp_description', ['venue' => $venue->name]),
            );
        }

        return [
            'status' => $created ? 'recorded' : 'already_recorded_today',
            'member' => $this->memberSummary($memberId),
            'venue' => [
                'id' => (int) $venue->id,
                'name' => (string) $venue->name,
            ],
            'visits_this_month' => $this->memberVisitCountThisMonth($memberId),
            'xp_awarded' => $engagement['xp_awarded'],
            'completed_challenges' => $engagement['completed_challenges'],
        ];
    }

    /**
     * The caller's own visit history.
     *
     * @return array<int, array<string, mixed>>
     */
    public function myVisits(int $userId, int $limit = 50): array
    {
        $tenantId = TenantContext::getId();

        return DB::table('partner_venue_visits as v')
            ->join('partner_venues as pv', function ($join) use ($tenantId): void {
                $join->on('pv.id', '=', 'v.venue_id')
                    ->where('pv.tenant_id', '=', $tenantId);
            })
            ->where('v.tenant_id', $tenantId)
            ->where('v.user_id', $userId)
            ->orderByDesc('v.visited_at')
            ->limit(max(1, min($limit, 200)))
            ->get(['v.id', 'v.visited_on', 'v.visited_at', 'pv.id as venue_id', 'pv.name as venue_name', 'pv.category'])
            ->map(static fn ($row): array => [
                'id' => (int) $row->id,
                'venue_id' => (int) $row->venue_id,
                'venue_name' => (string) $row->venue_name,
                'category' => $row->category,
                'visited_on' => (string) $row->visited_on,
                'visited_at' => $row->visited_at,
            ])
            ->all();
    }

    /**
     * Per-venue engagement rollup for tenant admins.
     *
     * @return array<string, mixed>
     */
    /**
     * NOTE ON SCOPE: total_visits/unique_members are deliberately LIFETIME
     * figures (the admin card reads "Total visits recorded"); only
     * recent_visits honours $days. So the grouped scan covers the tenant's
     * whole visit history by design, served by the
     * (tenant_id, venue_id, visited_at) index. If a tenant ever exceeds
     * roughly a million visits this wants a nightly rollup rather than a
     * live GROUP BY — it is not a bug, it is a known scale ceiling.
     */
    public function summary(int $days = 30): array
    {
        $tenantId = TenantContext::getId();
        $since = now()->subDays(max(1, min($days, 365)))->startOfDay();

        $rows = DB::table('partner_venue_visits as v')
            ->join('partner_venues as pv', function ($join) use ($tenantId): void {
                $join->on('pv.id', '=', 'v.venue_id')
                    ->where('pv.tenant_id', '=', $tenantId);
            })
            ->where('v.tenant_id', $tenantId)
            ->groupBy('pv.id', 'pv.name')
            ->selectRaw('pv.id AS venue_id, pv.name AS venue_name, COUNT(*) AS total_visits, COUNT(DISTINCT v.user_id) AS unique_members')
            ->selectRaw('SUM(CASE WHEN v.visited_at >= ? THEN 1 ELSE 0 END) AS recent_visits', [$since])
            ->orderByDesc('total_visits')
            ->get();

        return [
            'window_days' => $days,
            'total_visits' => (int) $rows->sum('total_visits'),
            'venues' => $rows->map(static fn ($row): array => [
                'venue_id' => (int) $row->venue_id,
                'venue_name' => (string) $row->venue_name,
                'total_visits' => (int) $row->total_visits,
                'unique_members' => (int) $row->unique_members,
                'recent_visits' => (int) $row->recent_visits,
            ])->all(),
        ];
    }

    /**
     * Visit rows for admin reporting / CSV export.
     *
     * @return array<int, array<string, mixed>>
     */
    public function visitRows(?int $venueId = null, ?string $from = null, ?string $to = null, int $limit = 5000): array
    {
        $tenantId = TenantContext::getId();

        return DB::table('partner_venue_visits as v')
            ->join('partner_venues as pv', function ($join) use ($tenantId): void {
                $join->on('pv.id', '=', 'v.venue_id')
                    ->where('pv.tenant_id', '=', $tenantId);
            })
            ->leftJoin('users as m', function ($join) use ($tenantId): void {
                $join->on('m.id', '=', 'v.user_id')
                    ->where('m.tenant_id', '=', $tenantId);
            })
            ->leftJoin('users as s', function ($join) use ($tenantId): void {
                $join->on('s.id', '=', 'v.recorded_by_user_id')
                    ->where('s.tenant_id', '=', $tenantId);
            })
            ->where('v.tenant_id', $tenantId)
            ->when($venueId !== null, fn ($query) => $query->where('v.venue_id', $venueId))
            ->when($from !== null, fn ($query) => $query->where('v.visited_on', '>=', $from))
            ->when($to !== null, fn ($query) => $query->where('v.visited_on', '<=', $to))
            ->orderByDesc('v.visited_at')
            ->limit(max(1, min($limit, 20000)))
            ->get([
                'v.id', 'v.visited_on', 'v.visited_at', 'v.source',
                'pv.name as venue_name',
                'v.user_id', 'm.first_name as member_first_name', 'm.last_name as member_last_name',
                's.first_name as staff_first_name', 's.last_name as staff_last_name',
            ])
            ->map(static fn ($row): array => [
                'id' => (int) $row->id,
                'visited_on' => (string) $row->visited_on,
                'visited_at' => $row->visited_at,
                'venue_name' => (string) $row->venue_name,
                'member_id' => (int) $row->user_id,
                'member_name' => trim(($row->member_first_name ?? '') . ' ' . ($row->member_last_name ?? '')),
                'recorded_by' => trim(($row->staff_first_name ?? '') . ' ' . ($row->staff_last_name ?? '')),
                'source' => (string) $row->source,
            ])
            ->all();
    }

    private function memberVisitCountThisMonth(int $userId): int
    {
        return PartnerVenueVisit::query()
            ->where('user_id', $userId)
            ->where('visited_on', '>=', now()->startOfMonth()->toDateString())
            ->count();
    }

    /**
     * @return array<string, mixed>
     */
    private function memberSummary(int $userId): array
    {
        $user = DB::table('users')
            ->where('id', $userId)
            ->where('tenant_id', TenantContext::getId())
            ->first(['id', 'first_name', 'last_name', 'avatar_url']);

        return [
            'id' => $userId,
            'name' => $user !== null
                ? trim(($user->first_name ?? '') . ' ' . ($user->last_name ?? ''))
                : '',
            'avatar_url' => $user->avatar_url ?? null,
        ];
    }
}

File: aspnet-backend/src/Nexus.Api/Services/PartnerVenueVisitService.cs
// Copyright (c) 2024-2026 Jasper Ford
// SPDX-License-Identifier: AGPL-3.0-or-later
// Author: Jasper Ford
// See NOTICE file for attribution and acknowledgements.

using System.Security.Cryptography;
using Microsoft.EntityFrameworkCore;
using Nexus.Api.Data;
using Nexus.Api.Entities;

namespace Nexus.Api.Services;

/// <summary>
/// Member pass issuing/rotation and staff-scanned visit recording — Laravel
/// parity for App\Services\PartnerVenueVisitService. The recordVisit rule
/// ladder is reproduced in the exact Laravel order, including the database
/// unique key treating a same-day rescan as already_recorded_today rather
/// than an error, and the self-attestation block.
/// </summary>
public class PartnerVenueVisitService
{
    public const string EngagementAction = "venue_visit";

    private readonly NexusDbContext _db;
    private readonly PartnerVenueService _venues;
    private readonly GamificationService _gamification;
    private readonly ChallengeService _challenges;
    private readonly ILogger<PartnerVenueVisitService> _logger;

    public PartnerVenueVisitService(
        NexusDbContext db,
        PartnerVenueService venues,
        GamificationService gamification,
        ChallengeService challenges,
        ILogger<PartnerVenueVisitService> logger)
    {
        _db = db;
        _venues = venues;
        _gamification = gamification;
        _challenges = challenges;
        _logger = logger;
    }

    public sealed record PassView(string Token, string QrUrl, string Status, string? LastUsedAt);

    /// <summary>
    /// One pass per (tenant, user), created lazily. A non-active pass found by
    /// its owner is flipped back to active. The unique index on
    /// (tenant_id, user_id) serialises the check-then-insert race: a losing
    /// writer re-reads the winner's row.
    /// </summary>
    public async Task<PassView> GetOrCreatePassAsync(
        int tenantId, int userId, string checkinBaseUrl, CancellationToken ct)
    {
        var pass = await FindPassAsync(tenantId, userId, ct);
        if (pass is null)
        {
            pass = new PartnerMemberPass
            {
                TenantId = tenantId,
                UserId = userId,
                Token = NewToken(),
                Status = PartnerMemberPass.StatusActive,
                CreatedAt = DateTime.UtcNow
            };
            _db.PartnerMemberPasses.Add(pass);
            try
            {
                await _db.SaveChangesAsync(ct);
            }
            catch (DbUpdateException)
            {
                _db.Entry(pass).State = EntityState.Detached;
                pass = await FindPassAsync(tenantId, userId, ct)
                    ?? throw new InvalidOperationException("Pass row vanished after unique-key race.");
            }
        }

        if (pass.Status != PartnerMemberPass.StatusActive)
        {
            pass.Status = PartnerMemberPass.StatusActive;
            pass.UpdatedAt = DateTime.UtcNow;
            await _db.SaveChangesAsync(ct);
        }

        return ToPassView(pass, checkinBaseUrl);
    }

    /// <summary>
    /// Rotates the token in place — same row, fresh 64-hex token, status
    /// forced active. The old token stops resolving immediately;
    /// last_used_at is deliberately NOT reset.
    /// </summary>
    public async Task<PassView> RotatePassAsync(
        int tenantId, int userId, string checkinBaseUrl, CancellationToken ct)
    {
        await GetOrCreatePassAsync(tenantId, userId, checkinBaseUrl, ct);
        var pass = await FindPassAsync(tenantId, userId, ct)
            ?? throw new InvalidOperationException("Pass row missing after get-or-create.");
        pass.Token = NewToken();
        pass.Status = PartnerMemberPass.StatusActive;
        pass.UpdatedAt = DateTime.UtcNow;
        await _db.SaveChangesAsync(ct);
        return ToPassView(pass, checkinBaseUrl);
    }

    public async Task<List<object>> MyVisitsAsync(int userId, int limit, CancellationToken ct)
    {
        limit = Math.Max(1, Math.Min(limit, 200));
        var rows = await _db.PartnerVenueVisits
            .AsNoTracking()
            .Where(v => v.UserId == userId)
            .Join(_db.PartnerVenues.IgnoreQueryFilters(), v => v.VenueId, venue => venue.Id,
                (v, venue) => new
                {
                    v.Id,
                    v.VenueId,
                    VenueName = venue.Name,
                    venue.Category,
                    v.VisitedOn,
                    v.VisitedAt
                })
            .OrderByDescending(row => row.VisitedAt)
            .Take(limit)
            .ToListAsync(ct);

        return rows.Select(row => (object)new
        {
            id = row.Id,
            venue_id = row.VenueId,
            venue_name = row.VenueName,
            category = row.Category,
            visited_on = row.VisitedOn.ToString("yyyy-MM-dd"),
            visited_at = row.VisitedAt
        }).ToList();
    }

    public sealed record RecordVisitResult(string Status, object? Payload);

    /// <summary>
    /// The Laravel rule ladder, in order: invalid pass → 404; scanner with no
    /// eligible venues, wrong venue, inactive venue, or self-scan → forbidden;
    /// several eligible venues and none chosen → needs_venue; then insert with
    /// the daily unique key deciding recorded vs already_recorded_today.
    /// XP and challenge progress move only on a genuinely new visit.
    /// </summary>
    public async Task<RecordVisitResult> RecordVisitAsync(
        int tenantId, string token, int staffUserId, int? venueId, CancellationToken ct)
    {
        var pass = await _db.PartnerMemberPasses
            .IgnoreQueryFilters()
            .FirstOrDefaultAsync(p => p.TenantId == tenantId
                && p.Token == token
                && p.Status == PartnerMemberPass.StatusActive, ct);
        if (pass is null) return new RecordVisitResult("invalid_pass", null);

        var memberId = pass.UserId;
        var eligible = await _venues.VenuesForStaffAsync(tenantId, staffUserId, ct);
        if (eligible.Count == 0) return new RecordVisitResult("forbidden", null);

        if (venueId.HasValue)
        {
            if (!eligible.Contains(venueId.Value)) return new RecordVisitResult("forbidden", null);
        }
        else if (eligible.Count == 1)
        {
            venueId = eligible[0];
        }
        else
        {
            var choices = await _db.PartnerVenues
                .AsNoTracking()
                .Where(v => eligible.Contains(v.Id))
                .OrderBy(v => v.Name)
                .Select(v => new { id = v.Id, name = v.Name })
                .ToListAsync(ct);
            return new RecordVisitResult("needs_venue", new
            {
                status = "needs_venue",
                venues = choices
            });
        }

        var venue = await _db.PartnerVenues
            .AsNoTracking()
            .FirstOrDefaultAsync(v => v.Id == venueId.Value, ct);
        if (venue is null || venue.Status != PartnerVenue.StatusActive)
        {
            return new RecordVisitResult("forbidden", null);
        }

        // A self-recorded visit would make the ledger self-attested.
        if (memberId == staffUserId) return new RecordVisitResult("forbidden", null);

        var now = DateTime.UtcNow;
        var today = DateOnly.FromDateTime(now);
        var created = true;
        var visit = new PartnerVenueVisit
        {
            TenantId = tenantId,
            VenueId = venue.Id,
            UserId = memberId,
            RecordedByUserId = staffUserId,
            Source = PartnerVenueVisit.SourceMemberPass,
            VisitedOn = today,
            VisitedAt = now,
            CreatedAt = now
        };
        _db.PartnerVenueVisits.Add(visit);
        try
        {
            await _db.SaveChangesAsync(ct);
        }
        catch (DbUpdateException)
        {
            // The daily unique key fired: same member, venue, and day.
            _db.Entry(visit).State = EntityState.Detached;
            created = false;
        }

        pass.LastUsedAt = now;
        pass.UpdatedAt = now;
        await _db.SaveChangesAsync(ct);

        var xpAwarded = 0;
        var completedChallenges = new List<object>();
        if (created)
        {
            try
            {
                var award = await _gamification.AwardXpAsync(
                    memberId, XpLog.Amounts.VenueVisit, XpLog.Sources.VenueVisit,
                    venue.Id, $"Visited {venue.Name}");
                if (award.Success) xpAwarded = XpLog.Amounts.VenueVisit;
                var completions = await _challenges.UpdateProgressAsync(memberId, EngagementAction);
                completedChallenges = completions
                    .Select(c => (object)new { id = c.Id, title = c.Title, xp_reward = c.XpReward })
                    .ToList();
            }
            catch (Exception ex)
            {
                // Engagement side effects must never fail the recorded visit.
                _logger.LogWarning(ex, "[PartnerVenues] engagement update failed for user {UserId}", memberId);
            }
        }

        var member = await _db.Users
            .IgnoreQueryFilters()
            .AsNoTracking()
            .Where(u => u.Id == memberId && u.TenantId == tenantId)
            .Select(u => new { u.Id, u.FirstName, u.LastName, u.AvatarUrl })
            .FirstOrDefaultAsync(ct);

        var monthStart = new DateOnly(today.Year, today.Month, 1);
        var visitsThisMonth = await _db.PartnerVenueVisits
            .IgnoreQueryFilters()
            .AsNoTracking()
            .CountAsync(v => v.TenantId == tenantId
                && v.UserId == memberId
                && v.VisitedOn >= monthStart, ct);

        return new RecordVisitResult(created ? "recorded" : "already_recorded_today", new
        {
            status = created ? "recorded" : "already_recorded_today",
            member = member is null ? null : new
            {
                id = member.Id,
                name = $"{member.FirstName} {member.LastName}".Trim(),
                avatar_url = member.AvatarUrl
            },
            venue = new { id = venue.Id, name = venue.Name },
            visits_this_month = visitsThisMonth,
            xp_awarded = xpAwarded,
            completed_challenges = completedChallenges
        });
    }

    /// <summary>
    /// Laravel summary(): total_visits and unique_members are LIFETIME; only
    /// recent_visits honours the window. window_days echoes the raw request
    /// value while the SQL clamps to 1..365. Ordered by total_visits desc.
    /// </summary>
    public async Task<object> SummaryAsync(int days, CancellationToken ct)
    {
        var clamped = Math.Max(1, Math.Min(days, 365));
        var since = DateTime.UtcNow.Date.AddDays(-clamped);

        var perVenue = await _db.PartnerVenueVisits
            .AsNoTracking()
            .GroupBy(v => v.VenueId)
            .Select(g => new
            {
                VenueId = g.Key,
                TotalVisits = g.Count(),
                UniqueMembers = g.Select(v => v.UserId).Distinct().Count(),
                RecentVisits = g.Count(v => v.VisitedAt >= since)
            })
            .ToListAsync(ct);

        var names = await _db.PartnerVenues
            .AsNoTracking()
            .Where(v => perVenue.Select(p => p.VenueId).Contains(v.Id))
            .ToDictionaryAsync(v => v.Id, v => v.Name, ct);

        var venues = perVenue
            .OrderByDescending(p => p.TotalVisits)
            .Select(p => new
            {
                venue_id = p.VenueId,
                venue_name = names.TryGetValue(p.VenueId, out var name) ? name : string.Empty,
                total_visits = p.TotalVisits,
                unique_members = p.UniqueMembers,
                recent_visits = p.RecentVisits
            })
            .ToList();

        return new
        {
            window_days = days,
            total_visits = venues.Sum(v => v.total_visits),
            venues
        };
    }

    public sealed record VisitRow(
        long Id, string VisitedOn, string VisitedAt, string VenueName,
        int MemberId, string MemberName, string RecordedBy, string Source);

    public async Task<List<VisitRow>> VisitRowsAsync(
        int? venueId, DateOnly? from, DateOnly? to, CancellationToken ct, int limit = 5000)
    {
        limit = Math.Max(1, Math.Min(limit, 20000));
        var query = _db.PartnerVenueVisits.AsNoTracking();
        if (venueId.HasValue) query = query.Where(v => v.VenueId == venueId.Value);
        if (from.HasValue) query = query.Where(v => v.VisitedOn >= from.Value);
        if (to.HasValue) query = query.Where(v => v.VisitedOn <= to.Value);

        var rows = await query
            .OrderByDescending(v => v.VisitedAt)
            .Take(limit)
            .Select(v => new
            {
                v.Id,
                v.VisitedOn,
                v.VisitedAt,
                VenueName = _db.PartnerVenues.IgnoreQueryFilters()
                    .Where(venue => venue.Id == v.VenueId)
                    .Select(venue => venue.Name)
                    .FirstOrDefault(),
                v.UserId,
                MemberFirst = _db.Users.IgnoreQueryFilters()
                    .Where(u => u.Id == v.UserId && u.TenantId == v.TenantId)
                    .Select(u => u.FirstName).FirstOrDefault(),
                MemberLast = _db.Users.IgnoreQueryFilters()
                    .Where(u => u.Id == v.UserId && u.TenantId == v.TenantId)
                    .Select(u => u.LastName).FirstOrDefault(),
                RecorderFirst = _db.Users.IgnoreQueryFilters()
                    .Where(u => u.Id == v.RecordedByUserId && u.TenantId == v.TenantId)
                    .Select(u => u.FirstName).FirstOrDefault(),
                RecorderLast = _db.Users.IgnoreQueryFilters()
                    .Where(u => u.Id == v.RecordedByUserId && u.TenantId == v.TenantId)
                    .Select(u => u.LastName).FirstOrDefault(),
                v.Source
            })
            .ToListAsync(ct);

        return rows.Select(r => new VisitRow(
            r.Id,
            r.VisitedOn.ToString("yyyy-MM-dd"),
            r.VisitedAt?.ToString("yyyy-MM-dd HH:mm:ss") ?? string.Empty,
            r.VenueName ?? string.Empty,
            r.UserId,
            $"{r.MemberFirst} {r.MemberLast}".Trim(),
            $"{r.RecorderFirst} {r.RecorderLast}".Trim(),
            r.Source)).ToList();
    }

    private async Task<PartnerMemberPass?> FindPassAsync(int tenantId, int userId, CancellationToken ct)
    {
        return await _db.PartnerMemberPasses
            .IgnoreQueryFilters()
            .FirstOrDefaultAsync(p => p.TenantId == tenantId && p.UserId == userId, ct);
    }

    private static PassView ToPassView(PartnerMemberPass pass, string checkinBaseUrl) => new(
        pass.Token,
        checkinBaseUrl.TrimEnd('/') + "/" + pass.Token,
        pass.Status,
        pass.LastUsedAt?.ToString("yyyy-MM-dd'T'HH:mm:ssK"));

    /// <summary>64 lowercase hex chars — bin2hex(random_bytes(32)) parity.</summary>
    private static string NewToken() =>
        Convert.ToHexString(RandomNumberGenerator.GetBytes(32)).ToLowerInvariant();
}
